Extensions/Screenshot/screenshottools.js

```diff
@@ -0,0 +1,29 @@
+/**
+ * @memberof gdjs
+ * @class screenshot
+ * @static
+ * @private
+ */
+
+gdjs.screenshot = {}
+
+/**
+ * Save a screenshot of the game.
+ * @param {string} savepath The path where to save the screenshot
+ */
+gdjs.screenshot.takeScreenshot = function (savePath, runtimeScene) {
+  let electron = gdjs.RuntimeGamePixiRenderer.getElectron();
+  let canvas = null;
+  let content = null;
+  let fileSystem = null;
+
+  if (electron) {
+    fileSystem = electron.remote.require("fs");
+    canvas = runtimeScene._renderer._pixiRenderer.view;
+    content = canvas.toDataURL('image/png').replace('data:image/png;base64,', '');
+    try {
+      fileSystem.writeFileSync(savePath, content, 'base64');
```

Comment:
Prefer asynchronous APIs if possible:

```
fs.writeFile(savePath, content, 'base64', (err) => {
  if (err) {
    console.error("Unable to save the screenshot at path: " + savePath);
  }
});
```

This is slightly less readable :'( (that's why Promise and async/await were invented - but async/await support is not widespread enough alas). But as it's asynchronous, you're not blocking the JavaScript thread, meaning that the game will be not, at least less, interrupted during the screenshot saving :)

Reply:
(you don't need the try catch with the asynchronous callback, it's handled by the `if (err) { .. }`)

Reply:
All I need is the canvas DOM element. The renderer, if PIXI or Cocos shouldn't matter at all. I wonder If we could expose the canvas one layer above so that I don't need to care about the renderer.

Reply:
Mmm sure we could expose the canvas one layer above, but on the other hand it's not even guaranteed that in the feature we won't have a renderer that is not HTML5 based. In fact, if you export with Cocos2d-JS for Android/iOS/desktop, the rendering is done natively with OpenGL, you don't even have the notion of a canvas element. (but I don't necessarily push this because HTML5 rendering with WebGL is as fast and more supported).

What we can do, is expose a "getCanvas" method in both Cocos2d-JS renderer and pixi renderer of runtimeGame, which can return null, and do this:

```
const canvas = runtimeScene.getGame().getRenderer().getCanvas();
if (canvas) {
   ...
} else {
  console.error("Screenshot are not supported on rendering engines without canvas.");
}
```

Knowing that `getCanvas` can return `null` if there is no canvas.

This way we are both compatible with Cocos2d-JS (HTML5 version) AND Pixi, and we still don't rely on internal values (instead we extend the renderer interface with a new function that they are supposed to implement).

Reply:
@Wend1go Let me know if something is not clear or if you don't see yourself doing that change.

Reply:
I'll look into it at the weekend. Caught a cold and my brain currently isn't working as expected.

Reply:
Sure no worries, take your time :)

Reply:
So I have added the getCanvas function to `runtimegame-pixi-renderer.js` and access it via `const canvas = runtimeScene.getGame().getRenderer().getCanvas();` which is working (still need to do it for cocos-js)
```
/**
 * Get the canvas DOM element.
 */
gdjs.RuntimeGamePixiRenderer.prototype.getCanvas = function() {
    return this._pixiRenderer.view;
}
```
But I noticed that I get the electron reference via `const electron = gdjs.RuntimeGamePixiRenderer.getElectron();` which is not "renderer neutral". I tried to get the electron reference via the `getElectron` function in `runtimegame-pixi-renderer.js` the same way as I get the canvas but the debugger tells me that it does not exist. This seems to be because it doesn't have the "prototype" prefix.
Can I somehow access the electron reference without going over `gdjs.RuntimeGamePixiRenderer.getElectron`?

Reply:
This is because `gdjs.RuntimeGamePixiRenderer.getElectron` is defined on the class directly, i.e: it's a "static method" (a method that does not need an instance of the class to run.

What you can do is:
* Convert it back to a method of the class by putting it on the prototype:
`gdjs.RuntimeGamePixiRenderer.prototype.getElectron = function() { ... }`
* Convert all the calls to it in the class from `gdjs.RuntimeGamePixiRenderer.getElectron()` to `this.getElectron()`
* Nothing is using getElectron outside of this class (otherwise, you would have needed to change the call to `runtimeScene.getGame().getRenderer().getElectron()`).

Finally, add `getElectron` to the Cocos2d-JS renderer (`gdjs.RuntimeGameCocosRenderer.prototype.getElectron`) and make the call to `runtimeScene.getGame().getRenderer().getElectron()` in your extension (which is now valid because "renderers" are required to implement getElectron).

Ideally, if we would have proper typing, both `RuntimeGamePixiRenderer` and `RuntimeGameCocosRenderer` would inherit from a common "interface" (also called "abstract class") which would require them to implement a set of method (getCanvas, getElectron).
As we don't have typing in JavaScript, we have to do it by ourselves (i.e: being vigilant).

Typing like Typescript or Flow would help but that's another story (which is why I'm progressively adding JSDoc types comment annotations, as they are understand by typescript and could in the future allow Typescript to check the typing of the whole codebase :)).

Reply:
The PixiJS renderer is working now. 
I added the code which I "think" is correct to the cocos renderer but I can't test this since I don't know how to package a test game into an electron App (with debugging tools enabled).

Reply:
I know that the way I access the canvas in cocosjs is foo bar. This is just for testing until I have found the proper way to get the canvas element.